---
title: "Heritability"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, 
                      message = FALSE, comment = '',
                      engine.opts = list(bash = "-l"))
options(width = 999)
```

```{r pkgs}
library(tidyverse)
library(SeqArray) 
library(SeqVarTools) 
library(SNPRelate)
library(GENESIS)
library(Biobase)
library(cowplot)
library(ggbeeswarm)
library(ggthemes)
```

## Sample:

- 445 individuals from The 1000 Genomes Project (358 from 4 European populations + 87 Yorubans) with RNA-seq available from the Geuvadis consortium.

## Data pre-processing

### Genotypes

We use VCF files provided by 1000 Genomes Phase III (low coverage, in GRCh38 coordinates).

We select the GEUVADIS individuals, and exclude variants with missing data or which are monomorphic in this subset of individuals.

Then we convert VCF to GDS, which is a data structure that optimizes working with genotype data in R.

To perform all this processing, we run the script `./launch_processVCF.sh`.

Then we concatenate the GDS files for each chromosome in a single GDS with the code below:

```{r VcfToGds, eval=FALSE}
gds_list <- sprintf("/raid/genevol/heritability/data/kgp/chr%d.gds", 1:22)

gds_file <- "/raid/genevol/heritability/data/kgp/allchrs.gds"

seqMerge(gds_list, gds_file)
```

We reduce the amount of variants by LD pruning:

```{r, eval=FALSE}
# this step takes hours
gds <- seqOpen(gds_file)

set.seed(100)
pruned <- snpgdsLDpruning(gds, 
                          method = "corr", 
                          ld.threshold = sqrt(0.1))

prunedsnps <- unlist(pruned, use.names = FALSE)

seqSetFilter(gds, variant.id = prunedsnps)
seqExport(gds, "/raid/genevol/heritability/data/kgp/allchrs_pruned.gds")
seqClose(gds)
```


### Phenotypes

We use the FASTQ files provided by the Geuvadis Consortium, and estimate expression with Salmon. We begin with the expression of HLA-A.

Transcript Per Million (TPM) values are quantile normal transformed by `QTLtools correct`. Here is a comparison between raw and standardize TPMs:


```{r}
# raw
hla_expression <- 
    "/raid/genevol/heritability/data/geuvadis_expression.bed.gz" %>%
    read_tsv() %>%
    filter(id == "HLA-A") %>%
    pivot_longer(-(1:6), names_to = "sampleid") %>%
    select(sampleid, value)

# standardized by QTLtools
hla_expression_std <- 
    "/raid/genevol/heritability/data/geuvadis_expression_std.bed" %>%
    read_tsv() %>%
    filter(id == "HLA-A") %>%
    pivot_longer(-(1:6), names_to = "sampleid") %>%
    select(sampleid, value)
```

```{r, echo = FALSE}
h1 <- ggplot(hla_expression, aes(value)) +
    geom_histogram() +
    labs(x = NULL, title = "raw")

h2 <- ggplot(hla_expression_std, aes(value)) +
    geom_histogram() +
    labs(x = NULL, title = "standardized")

s1 <- left_join(hla_expression, hla_expression_std, 
          by = "sampleid", suffix = c("_raw", "_std")) %>%
    ggplot(aes(value_raw, value_std)) +
    geom_point() +
    labs(x = "Raw", y = "Standardized")

plot_grid(h1, h2, s1, nrow = 2)
```

## Results

### PCA on genotype data
```{r}
# Close GDS and open the pruned one
gdsfmt::showfile.gds(closeall = TRUE)

gds <- seqOpen("/raid/genevol/heritability/data/kgp/allchrs_pruned.gds")

sample_ids <- seqGetData(gds, "sample.id")

pca <- snpgdsPCA(gds, num.thread = 16L)

pops <- read_tsv("/raid/genevol/heritability/data/geuvadis_metadata.tsv") %>%
    distinct(sampleid, population)

pcadf <- as.data.frame(pca$eigenvect) %>%
    as_tibble() %>%
    add_column(sampleid = pca$sample.id, .before = 1) %>%
    inner_join(pops, ., by = "sampleid")
```

```{r pcaplot, echo=FALSE}
p1 <- ggplot(pcadf, aes(V1, V2, color = population)) +
    geom_point() +
    scale_color_colorblind() +
    theme(panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "grey45"),
          legend.key = element_rect(fill = "grey45")) +
    labs(color = "Population")

p2 <- ggplot(pcadf, aes(V2, V3, color = population)) +
    geom_point() +
    scale_color_colorblind() +
    theme(panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "grey45"))

p3 <- ggplot(pcadf, aes(V3, V4, color = population)) +
    geom_point() +
    scale_color_colorblind() +
    theme(panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "grey45"))

p_legend <- get_legend(p1)

plot_grid(p1 + theme(legend.position = "none"), 
          p_legend,
          p2 + theme(legend.position = "none"), 
          p3 + theme(legend.position = "none"),
          nrow = 2)
```


### HLA-A expression across populations

```{r}
sample_info <- "/raid/genevol/heritability/data/geuvadis_metadata.tsv" %>%
    read_tsv() %>%
    select(-ena_id)

expression_df <- left_join(hla_expression_std, sample_info, by = "sampleid") %>%
    select(sampleid, population, sex, lab, value)

ggplot(expression_df, aes(reorder(population, value), value)) +
    geom_quasirandom(aes(color = population), method = "smiley") +
    geom_boxplot(fill = NA, outlier.color = NA) +
    scale_color_colorblind() +
    theme_bw() +
    labs(x = NULL, y = "Std TPM")
```

### HLA-A expression across laboratories

```{r}
ggplot(expression_df, aes(reorder(lab, value), value)) +
    geom_quasirandom(aes(color = lab), method = "smiley") +
    geom_boxplot(fill = NA, outlier.color = NA) +
    theme_bw() +
    labs(x = NULL, y = "Std TPM")
```

### HLA-A expression between males and females

```{r}
ggplot(expression_df, aes(reorder(sex, value), value)) +
    geom_quasirandom(aes(color = sex), method = "smiley") +
    geom_boxplot(fill = NA, outlier.color = NA) +
    theme_bw() +
    labs(x = NULL, y = "Std TPM")
```

Here we got to evaluate whether such differences across groupings are adequate for an analysis with the combined dataset, or whether they can be accounted for by using covariates.

Before we continue, we need to transform our phenotype data.frame into an "Annotated data frame", which is a data structure provided by `Biobase`. This means that we have to add metadata to our `expression_df` data.frame, and include a `sample.id` column as a requirement for `GENESIS`. Since we do not have multiple samples for the same individuals, we will just create the `sample.id` column as a copy of `sampleid`.


```{r}
metadata <- 
    c("sample identifier",
      "subject identifier",
      "laboratory of RNA sequencing",
      "subject's sex",
      "subject's population",
      "PC 1",
      "PC 2",
      "PC 3",
      "expression levels in TPM") %>%
    data.frame(labelDescription = .) 

annotphen <- expression_df %>%
    mutate(sample.id = sampleid) %>%
    left_join(select(pcadf, sampleid, V1:V3), by = "sampleid") %>%
    select(sample.id, sampleid, lab, sex, population, V1, V2, V3, value) %>%
    as.data.frame() %>%
    AnnotatedDataFrame(metadata)

# access the metadata with the varMetadata() function
varMetadata(annotphen)

# access the data with the pData() function
head(pData(annotphen))
```


## GRM

### GCTA

We will use the `SNPRelate` package to compute a GRM. We will begin with the GCTA method.

```{r}
# Close GDS and open the pruned one
gdsfmt::showfile.gds(closeall = TRUE)

pruned <- seqOpen("/raid/genevol/heritability/data/kgp/allchrs_pruned.gds")

# Computar a GRM
grm_obj <- snpgdsGRM(pruned, method = "GCTA", num.thread = 16L)
```

We extract and rename the matrix
```{r}
grm <- grm_obj$grm
rownames(grm) <- sample_ids
colnames(grm) <- sample_ids

# first 5 individuals:
grm[1:5, 1:5]
```

### Distribution of the GRM diagonal values
```{r, echo = FALSE}
grm_df <- grm %>%
    as.data.frame() %>%
    rownames_to_column("id1") %>%
    pivot_longer(-id1, names_to = "id2") %>%
    left_join(pops, by = c("id1" = "sampleid")) %>%
    left_join(pops, by = c("id2" = "sampleid")) %>%
    select(id1, id2, pop1 = population.x, pop2 = population.y, value)

grm_df %>%
    filter(id1 == id2) %>%
    ggplot(aes(reorder(id1, value), value)) +
    geom_bar(aes(fill = pop1, color = pop1), stat = "identity") +
    scale_fill_colorblind() +
    scale_color_colorblind() +
    geom_hline(yintercept = 1L, linetype = 2, size = 1, color = "grey") +
    theme_bw() +
    theme(panel.grid = element_blank(),
          axis.text.x = element_blank(),
          axis.line.x = element_blank(),
          axis.ticks.x = element_blank(),
          axis.title.x = element_blank()) +
    labs(color = "Population", fill = "Population",
         title = "GRM diagonal values") 
```

### GRM off-diagonal values

\* diagonal is set to NA so we can better see the range for off-diagonal (much smaller) values

```{r, echo = FALSE}
grm_offdiag <- grm_df %>%
    mutate(value = ifelse(id1 == id2, NA, value)) %>%
    arrange(pop1, pop2, id1, id2) %>%
    mutate(id1 = fct_inorder(id1),
           id2 = fct_inorder(id2))

lines_v <- grm_offdiag %>%
    group_by(pop1) %>%
    filter(1:n() == last(1:n())) %>%
    ungroup() %>%
    select(id1, pop1)

lines_h <- grm_offdiag %>%
    group_by(pop2) %>%
    filter(1:n() == last(1:n())) %>%
    ungroup() %>%
    select(id2, pop2)

labels_v <- grm_offdiag %>%
    group_by(pop1) %>%
    filter(1:n() == floor(median(1:n()))) %>%
    ungroup() %>%
    select(id1, pop1)

labels_h <- grm_offdiag %>%
    group_by(pop2) %>%
    filter(1:n() == floor(median(1:n()))) %>%
    ungroup() %>%
    select(id2, pop2)

ggplot(grm_offdiag) +
    geom_tile(aes(id1, id2, fill = value)) +
    scale_fill_gradient2(name = "GRM values", 
                         breaks = seq(round(min(grm_offdiag$value, na.rm = TRUE), 2),
                                      round(max(grm_offdiag$value, na.rm = TRUE), 2),
                                      by = 0.02)) +
    theme_bw() +
    theme(panel.grid = element_blank(),
          axis.text = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          plot.margin = unit(c(1,0,1,1), "lines")) +
    geom_hline(yintercept = lines_h$id2) +
    geom_vline(xintercept = lines_v$id1) +
    coord_cartesian(clip = "off") +
    geom_text(data = labels_v, aes(x = id1, y = 0, label = pop1), 
              vjust = 1.5) +
    geom_text(data = labels_h, aes(y = id2, x = 0, label = pop2), 
              vjust = -0.5, angle = 90) +
    labs(title = "Off-diagonal values")
```


### Fitting the Null model

The first step in finding genetic variants which are associated with a phenotype is preparing null model.

We fit a null model which adjusts gene expression according to covariates such as population of origin, laboratory of sequencing, and sex. We also account for the relatedness between individuals (GRM).

```{r}
mod_null <- fitNullModel(annotphen, 
                         outcome = "value", 
                         covars = c("lab", "sex", "V1", "V2"),
                         cov.mat = grm)
```

\*Note: if we include PC3, we get an error "the leading minor of order 445 is not positive definite"

### Association testing

Now that we have a Null model adjusting expression levels for sex, laboratory, population genetic structure, and relatedness, we can test for the association of the genetic variants with expression levels.

The first step is to create a `SeqVarData` object including both the GDS (genotypes) and the Annotated data.frame (phenotypes). 

Then we will use the `assocTestSingle` function to assess the effect of each variant.

```{r, assocTest}
# order individuals according to GDS
pData(annotphen) <- pData(annotphen) %>%
    mutate(sampleid = factor(sampleid, levels = seqGetData(pruned, "sample.id"))) %>%
    arrange(sampleid)

# create SeqVarData object and iterator
seqData <- SeqVarData(pruned, sampleData = annotphen)
iterator <- SeqVarBlockIterator(seqData, variantBlock = 20000, verbose = FALSE)

# test
assoc <- assocTestSingle(iterator, mod_null)

head(assoc, 10)
```


### Heritability

`GENESIS` includes the `varCompCI` to compute the proportion of variance explained (heritability) by each random effect, with a 95% CI: 

```{r}
varCompCI(mod_null, prop = TRUE)
```


## Package versions

```{r echo=FALSE}
devtools::session_info()
```
